File: typescript/src/advertisement/infrastructure/SqliteAdvertisementRepository.ts
import { AdvertisementRepository } from '../domain/AdvertisementRepository';
import { Advertisement } from '../domain/model/Advertisement';
import { DatabaseConnection } from '../../framework/database/DatabaseConnection';
import {Password} from "../domain/model/value-object/Password";

export class SqliteAdvertisementRepository implements AdvertisementRepository {

  constructor(
    private connection: DatabaseConnection) {
  }

  async findById(id: string): Promise<Advertisement> {

    const result = await this.connection.query(`SELECT * FROM advertisements WHERE id = ? `, [id])

    if (!result || result.length < 1) {
      throw new Error('Advertisement not found');
    }

    const row = result[0] as any;
    return new Advertisement(
      row.id,
      row.description,
      Password.fromEncryptedPassword(row.password),
      new Date(row.advertisement_date)
    )

  }

  async save(advertisement: Advertisement): Promise<void> {

    await this.connection.execute(
      `INSERT INTO advertisements (id, description, password, advertisement_date) 
      VALUES (?, ?, ?, ?) 
      ON CONFLICT(id) DO UPDATE 
      SET description = excluded.description, password = excluded.password, advertisement_date = excluded.advertisement_date`, [
      advertisement.id(),
      advertisement.description(),
      advertisement.password().value(),
      advertisement.date().toISOString(),
    ]);
  }
}

File: kotlin/src/main/kotlin/advertisement/infrastructure/persisence/SqLiteAdvertisementRepository.kt
package advertisement.infrastructure.persisence

import advertisement.domain.AdvertisementRepository
import advertisement.domain.model.Advertisement
import framework.database.DatabaseConnection
import java.security.MessageDigest
import java.time.LocalDateTime

class SqLiteAdvertisementRepository(private val connection: DatabaseConnection): AdvertisementRepository {
    override fun save(advertisement: Advertisement) {
        val passwordHash = advertisement.password!!.md5()
        connection.execute(
            "INSERT INTO advertisements (id, description, password, advertisement_date) VALUES ('" +
                    "${advertisement.id}', '${advertisement.description}', '$passwordHash', '${advertisement.date}') " +
                    "ON CONFLICT(id) DO UPDATE SET description = '${advertisement.description}', password = '$passwordHash', advertisement_date = '${advertisement.date}';"
        )
    }

    override fun findById(id: String): Advertisement {
        val result = connection.query(
            "SELECT * FROM advertisements WHERE id = '${id}'"
        )

        if (!result.next()) {
            throw Exception("Advertisement not found")
        }

        return Advertisement(
            result.getString("id"),
            result.getString("description"),
            result.getString("password"),
            LocalDateTime.parse(result.getString("advertisement_date")),
        )
    }

    private fun String.md5(): String {
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(this.toByteArray())
        return digest.toString()
    }
}
